controllers/k8ssandra/medusa_reconciler.go

```diff
@@ -0,0 +1,239 @@
+package k8ssandra
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	medusaapi "github.com/k8ssandra/k8ssandra-operator/apis/medusa/v1alpha1"
+	cassandra "github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/secret"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Generate a secret for Medusa or use the existing one if provided in the spec
+func (r *K8ssandraClusterReconciler) reconcileMedusaSecrets(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	logger logr.Logger,
+) error {
+	logger.Info("Reconciling Medusa user secrets")
+	if kc.Spec.Medusa != nil {
+		cassandraUserSecretRef := kc.Spec.Medusa.CassandraUserSecretRef
+		if cassandraUserSecretRef == "" {
+			cassandraUserSecretRef = cassandraUserSecretName(kc.Spec.Medusa, kc.Name)
+		}
+		logger = logger.WithValues(
+			"MedusaCassandraUserSecretRef",
+			cassandraUserSecretRef,
+		)
+		if err := secret.ReconcileSecret(ctx, r.Client, cassandraUserSecretRef, kc.Name, kc.Namespace); err != nil {
+			logger.Error(err, "Failed to reconcile Medusa CQL user secret")
+			return err
+		}
+	}
+	logger.Info("Medusa user secrets successfully reconciled")
+	return nil
+}
+
+// Get the cassandra user secret name from the medusa spec or generate one based on the cluster name
+func cassandraUserSecretName(medusaSpec *medusaapi.MedusaClusterTemplate, clusterName string) string {
+	if medusaSpec.CassandraUserSecretRef == "" {
+		return fmt.Sprintf("%s-medusa", clusterName)
+	}
+	return medusaSpec.CassandraUserSecretRef
+}
+
+// Create all things Medusa related in the cassdc podTemplateSpec
+func AddMedusa(dcConfig *cassandra.DatacenterConfig, medusaSpec *medusaapi.MedusaClusterTemplate, logger logr.Logger) error {
+	// Medusa isn't enabled (anymore) in this cluster
+	if medusaSpec != nil {
+		logger.Info("Medusa is enabled")
+		updateMedusaInitContainer(dcConfig, medusaSpec, logger)
+		updateMedusaMainContainer(dcConfig, medusaSpec, logger)
+		updateMedusaVolumes(dcConfig, medusaSpec, logger)
+	} else {
+		logger.Info("Medusa is not enabled")
+	}
+
+	return nil
+}
+
+func updateMedusaInitContainer(dcConfig *cassandra.DatacenterConfig, medusaSpec *medusaapi.MedusaClusterTemplate, logger logr.Logger) {
+	restoreContainerIndex, found := findMedusaContainer(dcConfig.PodTemplateSpec, "medusa-restore", true)
+	restoreContainer := &corev1.Container{Name: "medusa-restore"}
+	if found {
+		logger.Info("Found medusa-restore init container")
+		// medusa-restore init container already exists, we may need to update it
+		restoreContainer = dcConfig.PodTemplateSpec.Spec.InitContainers[restoreContainerIndex].DeepCopy()
+	}
+	restoreContainer.Image = fmt.Sprintf("%s/%s:%s", *medusaSpec.Image.Registry, medusaSpec.Image.Repository, *medusaSpec.Image.Tag)
```

Comment:
1. See my PR #243 : I think this `Sprintf` call is broken: the generated string contains a registry and a repository, but no image name, e.g. `docker.io/thelastpickle:latest`.
2. How are you handling default values? E.g. if the registry and the repository are both nils you'd get a nil pointer dereference here. And if they are set to their zero values (the empty string), you'd get something like `"/:"`. You probably need a function to generate the image string.

Reply:
Using the new stuff you created to deal with images now that I've rebased 👍